### PR TITLE
Add uniswapv2

### PR DESCRIPTION
**Pull request by MikeHathaway**

### CODE REVIEW

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+        register_keys(self.web3, self.arguments.eth_key)
+
+        # Record if eth is in pair, so can check which liquidity method needs to be used
+        self.is_eth = 'ETH' in self.pair()
+
+        # Assume token b is always ETH if ETH is in pair
+        token_a_name = self.pair().split('-')[0]
+        token_b_name = 'WETH' if self.pair().split('-')[1] == 'ETH' or self.pair().split('-')[1] == 'WETH' else self.pair().split('-')[1]
+
+        self.reloadable_config = ReloadableConfig(self.arguments.token_config)
```
Comment (grandizzy) [inline market_maker_keeper/uniswapv2_market_maker_keeper.py:110]: I don't think token config should be passed as an argument but since the config json is bundled within keeper code it should read directly from that file.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+                            help="Percentage difference between future Uniswap exchange rate and aggregated price"
+                                 "(default: 5)")
+
+        parser.add_argument("--debug", dest='debug', action='store_true',
+                            help="Enable debug output")
+
+        self.arguments = parser.parse_args(args)
+        setup_logging(self.arguments)
+
+        self.web3 = kwargs['web3'] if 'web3' in kwargs else Web3(HTTPProvider(endpoint_uri=f"{self.arguments.rpc_host}:{self.arguments.rpc_port}",
```
Comment (grandizzy) [inline market_maker_keeper/uniswapv2_market_maker_keeper.py:97]: can we add support for https endpoints too?

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+        assert (isinstance(uniswap_current_exchange_price, Wad))
+        assert (isinstance(accepted_slippage, Wad))
+
+        # Subtract Wad.from_number(1) to leave eth for gas
+        eth_desired = eth_balance - Wad.from_number(1.0)
+        if eth_desired < Wad.from_number(0):
+            self.logger.info(f"Insufficient Eth balance.")
+            return
+
+        # if self.uniswap.is_new_pool:
```
Comment (grandizzy) [inline market_maker_keeper/uniswapv2_market_maker_keeper.py:171]: should this logic be removed?

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+            self.logger.info(f"No liquidity to remove")
+
+    def place_liquidity(self):
+        """
+        Add or remove liquidity depending upon the difference between Uniswap ratio and our external price feeds.
+
+        Pricing is expressed as a function of token_a balance / token_b balance
+        """
+
+        # feed_price = self.feed.get()[0]['price']
```
Comment (grandizzy) [inline market_maker_keeper/uniswapv2_market_maker_keeper.py:395]: shouldn't this read from price feed instead using hardcodded value?

File: market_maker_keeper/uniswapv2_util.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
+#
```
Comment (grandizzy) [inline market_maker_keeper/uniswapv2_util.py:1]: support for UniswapV2 contract is better suitable to stay in pyexchange project . Abi file can be a separate one as https://github.com/makerdao/pyexchange/tree/master/pyexchange/abi

File: market_maker_keeper/uniswapv2_util.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+                uni_df.loc[:,"Input_2nd"] = new[1]
+                uni_df.loc[:,'readable_gas'] = uni_df['gas'].apply(self.to_int)
+                uni_df.loc[:,'readable_gasPrice'] = uni_df['gasPrice'].apply(self.to_int)
+                uni_df.loc[:,'readable_value'] = uni_df['value'].apply(self.to_int).astype(float)
+                #printing all pending transactions that are being sent to any uniswap exchange address
+                self.logger.debug(uni_df[["to","from","readable_gas","readable_gasPrice","readable_input","Input_1st","Input_2nd","readable_value","hash"]])
+                #here we are pruning the dataframe to only pending being sent to the DAI-ETH uniswap exchange contract
+                to_df = uni_df[uni_df["to"]==self.contract_address_lower]
+
+                self.logger.info(f'total number of ETH being sent in and sold in a token to token swap '
```
Comment (grandizzy) [inline market_maker_keeper/uniswapv2_util.py:267]: was under the impression it can cope with any pair tokens in UniswapV2 but I see logs pointing to ETH and DAI only

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 3 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:171]: Yes, I will remove it

File: market_maker_keeper/uniswapv2_util.py (hunk 6 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_util.py:267]: That file is no longer used. After talking with Akash, we are no longer going to attempt to predict future price moves based upon pending txs. I will remove it accordingly

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 2 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:97]: I believe this should already support https. I've been passing https endpoints to it

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 1 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:110]: That makes sense, I will adjust accordingly.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 4 above)
Comment (ith-harvey) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:395]: yes, would like to echo this. In order for Uniswap to work with our system the keeper will need to have some logic that handles our typical price feed input arg `--price-feed` not quite sure how `--uniswap-feed` works / will be different.

Most importantly, and you may already have this covered Mike (but didn't see anything in the below code), We need to be sure that if the price feed is giving Null values the bot stops placing orders and exits all liquidity for that pair

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+        parser.add_argument("--gas-price", type=int, default=9000000000,
+                            help="Gas price (in Wei)")
+
+        parser.add_argument("--smart-gas-price", dest='smart_gas_price', action='store_true',
+                            help="Use smart gas pricing strategy, based on the ethgasstation.info feed")
+
+        parser.add_argument("--initial-exchange-rate", type=float, default=None,
+                            help="Used to determine the initial ratio to be used in a newly created pool")
+
+        parser.add_argument("--percentage-difference", type=float, default=2,
```
Comment (ith-harvey) [inline market_maker_keeper/uniswapv2_market_maker_keeper.py:83]: what do you think about renaming this to accepted slipage percentage?

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+                            help="Percentage difference between Uniswap exchange rate and aggregated price"
+                                 "(default: 2)")
+
+        parser.add_argument("--uniswap-percentage-difference", type=float, default=5,
```
Comment (ith-harvey) [inline market_maker_keeper/uniswapv2_market_maker_keeper.py:87]: not quite sure what this one does... if I understand correctly we should just need accepted slippage.... could you unpack this one for me @MikeHathaway ?

File: market_maker_keeper/uniswapv2_util.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+import pandas as pd
+from web3 import Web3
+from time import gmtime, strftime
+import calendar
+import json
+import ast
+import logging
+
+uniswap_abi = [{"name": "NewExchange", "inputs": [{"type": "address", "name": "token", "indexed": True}, {"type": "address", "name": "exchange", "indexed": True}], "anonymous": False, "type": "event"}]
```
Comment (ith-harvey) [inline market_maker_keeper/uniswapv2_util.py:26]: What do you think about breaking out the ABI like we do in pymaker into its own file?

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 1 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:110]: Yes, the goal here is to provide generalized support for `Token` class. This config file can then be reused across different keepers, which would have let us simplify DyDx as just one example.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 7 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:83]: That sounds like a better name, renaming accordingly

File: market_maker_keeper/uniswapv2_util.py (hunk 9 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_util.py:26]: Yeah that's the approach taken elsewhere. This file was unneeded, so removed entirely.

File: market_maker_keeper/uniswapv2_util.py (hunk 5 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_util.py]: This has been removed

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 8 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py]: This is older code that was used to predict future price movements based on the mempool. I've removed it as its now unused.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 4 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py]: Added price feed support. Also, I've added logic on line 414 in `determine_liquidity_action()` that will remove all liquidity on a null price feed.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 10]
```diff
@@ -0,0 +1,441 @@
...
+        self.web3 = kwargs['web3'] if 'web3' in kwargs else Web3(HTTPProvider(endpoint_uri=f"{self.arguments.rpc_host}:{self.arguments.rpc_port}",
+                                                                              request_kwargs={"timeout": self.arguments.rpc_timeout}))
+
+        self.web3.eth.defaultAccount = self.arguments.eth_from
+        register_keys(self.web3, self.arguments.eth_key)
+
+        # Record if eth is in pair, so can check which liquidity method needs to be used
+        self.is_eth = 'ETH' in self.pair()
+
+        # Assume token b is always ETH if ETH is in pair
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:113]: yes, it works fine, and is expected. `token_b` can be any other token as well. Assuming `token_b` will always be Eth enabled me to simplify the logic in the eth related functions

I've tested this out with 7 different token pairings, with various decimals on Kovan and mainnet.

File: market_maker_keeper/reloadable_config.py [hunk 11]
```diff
@@ -43,6 +43,7 @@ def __init__(self, filename: str):
         assert(isinstance(filename, str))
 
         self.filename = filename
+        self._checksum = None
```
Comment (EdNoepel) [inline market_maker_keeper/reloadable_config.py:46]: Perhaps this should be named to indicate it's the checksum only for token configuration.

File: test.sh [hunk 12]
```diff
@@ -1,4 +1,20 @@
 #!/bin/sh
 source _virtualenv/bin/activate
 
-PYTHONPATH=$PYTHONPATH:./lib/pymaker:./lib/pyexchange:./lib/ethgasstation-client:./lib/gdax-client py.test --cov=market_maker_keeper --cov-report=term --cov-append tests/
+# Pull the docker image
+docker pull makerdao/testchain-pymaker:unit-testing
```
Comment (EdNoepel) [inline test.sh:5]: Again, why do we need the MCD Parity container?

File: docker-compose.yml [hunk 13]
```diff
@@ -17,4 +17,31 @@ services:
       --coinbase-password $${COINBASE_PASSWORD} \
       --price-feed $${ETH_DAI_PRICE_FEED?:} \
       --pair ETH-DAI \
-      --config /opt/maker/config/$${COINBASE_ETH_DAI_BANDS?:}"
\ No newline at end of file
+      --config /opt/maker/config/$${COINBASE_ETH_DAI_BANDS?:}"
+
+  parity:
```
Comment (EdNoepel) [inline docker-compose.yml:22]: I see the ganache image is used for unit testing.  Why do we need the MCD dockerized Parity testchain?

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+        if self.arguments.factory_address is None and self.arguments.router_address is None:
+            self.uniswap = UniswapV2(self.web3, self.token_a, self.token_b)
+        else:
+            self.uniswap = UniswapV2(self.web3, self.token_a, self.token_b, Address(self.arguments.router_address), Address(self.arguments.factory_address))
+
+    def main(self):
+        with Lifecycle(self.web3) as lifecycle:
+            lifecycle.initial_delay(self.arguments.initial_delay)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(10, self.place_liquidity)
```
Comment (EdNoepel) [inline market_maker_keeper/uniswapv2_market_maker_keeper.py:144]: I believe we'd usually have an `on_shutdown` handler to cancel open orders, here I suppose that would remove liquidity from the pool.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+
+    def startup(self):
+        self.uniswap.approve(self.token_a)
```
Comment (EdNoepel) [inline market_maker_keeper/uniswapv2_market_maker_keeper.py:147]: The PR with `pyexchange` implmentation (https://github.com/makerdao/pyexchange/pull/151/files#diff-8f7634f1e582be293e2d361a03d968caR117) uses `directly` for the two approval calls.  It'd be nice if an onchain call was added to check whether the approval needs to be performed before sending a TX.  Would save some gas each time the keeper was started.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+        self.token_a = list(filter(lambda token: token.name == token_a_name, token_config))[0]
+        self.token_b = list(filter(lambda token: token.name == token_b_name, token_config))[0]
+
+        self.gas_price = GasPriceFactory().create_gas_price(self.arguments)
+
+        self.accepted_slippage = Wad.from_number(self.arguments.accepted_slippage / 100)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
+        self.initial_exchange_rate = self.arguments.initial_exchange_rate
+        self.uniswap_current_exchange_price = Wad.from_number(self.initial_exchange_rate) if self.initial_exchange_rate is not None else self.uniswap.get_exchange_rate()
+        self.testing_feed_price = False
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:131]: Yes, this is still needed. I'm testing with a fixed price feed, but I needed it in order to be able to test the behavior of the keeper in the event of various price feed changes (i.e. price feed becoming null; price shifting enough to trigger remove liquidity, etc.)

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+        Pricing is expressed as a function of token_a balance / token_b balance in the contract
+
+        First calculate the acceptable price movement limit based upon the
+        difference between external price feeds and the accepted slippage.
+
+        If the accepted difference is greater than the distance of uniswaps price from external prices
+        add liquidity to the pool, otherwise remove it.
+        """
+
+        self.feed_price = self.price_feed.get_price().buy_price if self.testing_feed_price is False else self.test_price
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:335]: Yes, answered above

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+        self.token_b = list(filter(lambda token: token.name == token_b_name, token_config))[0]
+
+        self.gas_price = GasPriceFactory().create_gas_price(self.arguments)
+
+        self.accepted_slippage = Wad.from_number(self.arguments.accepted_slippage / 100)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
+        self.initial_exchange_rate = self.arguments.initial_exchange_rate
+        self.uniswap_current_exchange_price = Wad.from_number(self.initial_exchange_rate) if self.initial_exchange_rate is not None else self.uniswap.get_exchange_rate()
+        self.testing_feed_price = False
+        self.test_price = Wad.from_number(0)
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:132]: Yes, answered above

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+        Given available token balances and an exchange rate, calculations of amounts to add,
+        and limits for price movement need to be calculated.
+
+        Args:
+            uniswap_current_exchange_price: The current Product of the pool
+        Returns:
+            A Pymaker Receipt object or a None
+        """
+
+        token_a_balance = self.uniswap.get_account_token_balance(self.token_a) if not self.is_eth else self.uniswap.get_account_eth_balance()
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:208]: Yes, that was old code that has since been removed. I realized logic could be significantly simpler if Eth is assumed to be the base token. With this assumption, I can avoid a ton of logical complexity each time I query to check if the token is eth or not.

File: market_maker_keeper/reloadable_config.py (hunk 11 above)
Comment (MikeHathaway) [reply market_maker_keeper/reloadable_config.py:46]: That makes sense, I will update the name.

File: test.sh (hunk 12 above)
Comment (MikeHathaway) [reply test.sh:5]: Good catch, we don't. This was Copypasta. I've removed references to the parity container accordingly.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 15 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:147]: So, I can add comments around the approval, but that `directly` approve function will actually already check the allowance for a given token and contract prior to sending the approve transaction again unnecessarily: https://github.com/makerdao/pymaker/blob/master/pymaker/approval.py#L38

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 14 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:144]: Yes, good catch. I am working on adding this now

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+
+        """
+
+        exchange_token_a_balance = Wad.from_number(0) if self.uniswap.is_new_pool else self.uniswap.get_exchange_balance(self.token_a, self.uniswap.pair_address)
+        exchange_token_b_balance = Wad.from_number(0) if self.uniswap.is_new_pool else self.uniswap.get_exchange_balance(self.token_b, self.uniswap.pair_address)
+
+        self.logger.info(f"Exchange Contract {self.token_a.name} amount: {exchange_token_a_balance}; "
+                         f"Exchange Contract {self.token_b.name} amount: {exchange_token_b_balance}")
+
+        add_liquidity, remove_liquidity = self.determine_liquidity_action()
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:365]: We can add an assert, but I don't think this is necessary - As written it's not currently logically possible for them to both be true simultaneously. `diff` would have to be simultaneously be both smaller and larger than `feed_price - uniswap_current_exchange_price`. If the underlying numbers are None, it will trigger the `if case` and `remove_liquidity` would be hardcoded `True`, and `add_liquidity` would be `False`.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 21]
```diff
@@ -0,0 +1,379 @@
...
+        token_config = self.get_token_config().tokens
+        self.token_a = list(filter(lambda token: token.name == token_a_name, token_config))[0]
+        self.token_b = list(filter(lambda token: token.name == token_b_name, token_config))[0]
+
+        self.gas_price = GasPriceFactory().create_gas_price(self.arguments)
+
+        self.accepted_slippage = Wad.from_number(self.arguments.accepted_slippage / 100)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
+        self.initial_exchange_rate = self.arguments.initial_exchange_rate
+        self.uniswap_current_exchange_price = Wad.from_number(self.initial_exchange_rate) if self.initial_exchange_rate is not None else self.uniswap.get_exchange_rate()
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:130]: My thinking about having a seperate variable was that this would enable us to create a market for low liquidity tokens that had highly volatile or limited external price feeds, or tokens in which we had an explicit `product` or price we were trying to hit. While the product can be anything, once the product is set on creation, it exists at the same product in perpetuity.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 22]
```diff
@@ -0,0 +1,379 @@
...
+        self.testing_feed_price = False
+        self.test_price = Wad.from_number(0)
+
+        # used for local testing
+        if self.arguments.factory_address is None and self.arguments.router_address is None:
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:135]: Well we do pass in the args when we are running the integration tests in the pipeline and router / factory have been locally deployed. If local testing is being conducted the keeper will utilize the local contracts, otherwise it falls back on the default values which are the same for every pair and network. This can be changed of course, but what would the benefits be?

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:365]: Given simplicity and minimal cost, i've added an assert to check that the values aren't equal.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py [hunk 23]
```diff
@@ -0,0 +1,379 @@
...
+        self._last_config_dict = None
+        self._last_config = None
+        token_config = self.get_token_config().tokens
+        self.token_a = list(filter(lambda token: token.name == token_a_name, token_config))[0]
+        self.token_b = list(filter(lambda token: token.name == token_b_name, token_config))[0]
+
+        self.gas_price = GasPriceFactory().create_gas_price(self.arguments)
+
+        self.accepted_slippage = Wad.from_number(self.arguments.accepted_slippage / 100)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
```
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:130]: That makes sense, I've added control feed to the state machine determining whether or not we should add or remove liquidity.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 14 above)
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:172]: Shutdown handler has been added via `self._should_shutdown`, along with integration test `test_should_remove_liquidity_if_shutdown_signal_received`

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:154]: The address should always be the same for the router, it only varies frequently in the local testing setup, which is why I figured it could be simplified by hardcoding.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ith-harvey) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:151]: @MikeHathaway This is really interesting... I think for testing purposes it would be very cool to have a `disaster price feed` where it switches between null and crazy price movements to test bot edge cases... 

Sounds good, we will keep it here then.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (ith-harvey) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:147]: @MikeHathaway Lets discuss this with @mitakash today... 

My feeling is that having a price feed (for any token that we trade) whether it be illiquid or a shit token should have a price feed. Only for the rare exception of a stablecoin to stable coin pair should we have a fixed feed and even if that scenario took place we still have much more price protection when creating a stable coin price feed with a max price / min price then we ever would with fixed...

The above would take care of the fixed price idea, which is very valid. @MikeHathaway could you please describe what you mean by product in this context?

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (ith-harvey) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:154]: The address is always the same for Oasis contract (aside from it changing every year) The same goes for 0x smart contracts... I think we shouldn't have it hard coded

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (ith-harvey) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:447]: Thank you, always best just in case. Yes, I agree with you I don't see the logical possibility but best to protect us from those 
crazy scenarios. Just my opinion.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py:160]: I will update it to rely on the passed in address

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py]: That assumption has been removed, and ETH can now be either the base, or the quote currency.

File: market_maker_keeper/uniswapv2_market_maker_keeper.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (MikeHathaway) [reply market_maker_keeper/uniswapv2_market_maker_keeper.py]: Added a fallback to use our price feed in the event that a pool doesn't exist yet, as suggested. `initial_exchange_rate` has now been removed entirely.
